content_handler/flutter_runner/application.cc

```diff
@@ -0,0 +1,238 @@
+
+  settings_.script_snapshot_path = "snapshot_blob.bin";
+
+  settings_.log_tag = debug_label_ + std::string{"(flutter)"};
+
+  settings_.task_observer_add =
+      std::bind(&CurrentMessageLoopAddAfterTaskObserver, std::placeholders::_1,
+                std::placeholders::_2);
+
+  settings_.task_observer_remove = std::bind(
+      &CurrentMessageLoopRemoveAfterTaskObserver, std::placeholders::_1);
+}
+
+Application::~Application() = default;
+
+// |app::ApplicationController|
+void Application::Kill() {
+  if (last_return_code_.first) {
+    for (auto wait_callback : wait_callbacks_) {
+      wait_callback(last_return_code_.second);
+    }
+  }
+  wait_callbacks_.clear();
+
+  delegate_.OnApplicationTerminate(this);
+  // WARNING: Don't do anything past this point as this instance may have been
+  // collected.
+}
+
+// |app::ApplicationController|
+void Application::Detach() {
+  application_controller_.set_error_handler(nullptr);
+}
+
+// |app::ApplicationController|
+void Application::Wait(const WaitCallback& callback) {
+  wait_callbacks_.emplace_back(std::move(callback));
+}
+
+// |flutter::Engine::Delegate|
+void Application::OnEngineTerminate(const Engine* shell_holder) {
+  auto found = std::find_if(shell_holders_.begin(), shell_holders_.end(),
+                            [shell_holder](const auto& holder) {
+                              return holder.get() == shell_holder;
+                            });
+
+  if (found == shell_holders_.end()) {
+    return;
+  }
+
+  // We may launch multiple shell in this application. However, we will
+  // terminate when the last shell goes away. The error code return to the
+  // application controller will be the last isolate that had an error.
+  auto return_code = shell_holder->GetEngineReturnCode();
+  if (return_code.first) {
+    last_return_code_ = return_code;
+  }
+
+  shell_holders_.erase(found);
+
+  if (shell_holders_.size() == 0) {
+    Kill();
+    // WARNING: Don't do anything past this point because the delegate may have
+    // collected this instance via the termination callback.
+  }
+}
+
+void Application::CreateShellForView(
+    f1dl::InterfaceRequest<mozart::ViewProvider> view_provider_request) {
+  shells_bindings_.AddBinding(this, std::move(view_provider_request));
+}
+
+// |mozart::ViewProvider|
+void Application::CreateView(
+    f1dl::InterfaceRequest<mozart::ViewOwner> view_owner,
+    f1dl::InterfaceRequest<app::ServiceProvider>) {
+  if (!application_context_) {
+    FXL_DLOG(ERROR) << "Application context was invalid when attempting to "
+                       "create a shell for a view provider request.";
+    return;
+  }
+
+  // This method may be called multiple times. Care must be taken to ensure that
+  // all arguments can be accessed or synthesized multiple times.
+  // TODO(chinmaygarde): Figure out how to re-create the outgoing service
+  // request handle.
+  shell_holders_.emplace(std::make_unique<Engine>(
```

Comment:
So an isolate is created for each call to CreateView rather than each call to StartApplication?

Reply:
Yes, a shell is created per application, but a root isolate (UI) with window bindings is created per view.